File: py/3493_Properties_Graph.py
class Solution:
    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def unite(self, x, y):
        xp = self.find(x)
        yp = self.find(y)
        if xp == yp:
            return
        self.parent[yp] = xp
    
    def numberOfComponents(self, properties, k: int) -> int:
        n = len(properties)
        
        self.parent = list(range(n))
        
        freqs = [set(props) for props in properties]
        
        for i in range(n):
            for j in range(i+1, n):
                common = len(freqs[i] & freqs[j])
                if common >= k:
                    self.unite(i, j)
        
        components = set(self.find(i) for i in range(n))
        return len(components)

File: cpp/3493_Properties_Graph.cpp
class Solution {
public:
    vector<int> parent;
    int find(int x) {
        if (parent[x] == x) return x;
        return parent[x] = find(parent[x]);
    }
    void unite(int x, int y) {
        int xp = find(x);
        int yp = find(y);

        if (xp == yp) return;
        parent[yp] = xp;
    }
    int numberOfComponents(vector<vector<int>>& properties, int k) {
        int n = properties.size();
        vector<bitset<101>> freqs(n);  
             
        parent.resize(n);
        iota(parent.begin(), parent.end(), 0);

        for (int i = 0; i < n; ++i) {
            for (int p : properties[i])
                freqs[i][p] = 1;
        }

        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                int common = (freqs[i] & freqs[j]).count();
                if (common >= k)
                    unite(i, j);
            }
        }

        unordered_set<int> components;
        for (int i = 0; i < n; ++i)
            components.insert(find(i));
        return components.size();
    }
};

